content/blog/2019-10-23-posh/index.md

```diff
@@ -0,0 +1,315 @@
+        acct->balance -= amt; return true;
+    } else { return false; }
+  }
+}
+```
+where `atomic` was a hardware-supported feature for ensuring the atomicity of the contained code.
+If any other thread modified `acct->balance` during the execution of this transaction,
+it would *abort* and have to be retried or cancelled.
+
+Usually, HTM is implemented by piggy-backing off of the cache coherence protocol,
+which normally ensures that memory writes to the same address are eventually propagated
+between cores. Unfortunately, cache coherency can be [notoriously complex](https://doi.org/10.1109/2.55497),
+especially in the face of ambiguously defined and/or weak memory models.
+One might reasonably expect adding new synchronization features to introduce bugs
+and/or interact unexpectedly with existing weak memory guarantees.
+Furthermore, relying on cache coherency drastically limits size of datasets read or written by hardware transactions;
+in [most systems](https://researcher.watson.ibm.com/researcher/files/us-rodaira/ISCA2015_ComparisonOfHTM.pdf) the write set must fit entirely inside the L1 cache.
+
+
+### HTM Today
+
+In reality, hardware transactional memory has primarily been a failure
+and does not see wide use today.
+While Intel theoretically supports these kinds of instructions
+with [TSX](https://en.wikipedia.org/wiki/Transactional_Synchronization_Extensions),
+numerous bug reports have caused them to [disable it on a number of processors](https://www.anandtech.com/show/8376/intel-disables-tsx-instructions-erratum-found-in-haswell-haswelleep-broadwelly).
+Furthermore, the [limitations of TSX](https://blog.ret2.io/2019/06/26/attacking-intel-tsx/)
+and other such extensions often make using them impractical, unstable and/or insecure.
+
+However, some low-level code *does* utilize HTM to implement
+efficient libraries for high performance computing. In these instances,
+developers are targeting very specific architectures with very detailed
+models of the processor and memory systems. Since developers in
+this domain are already concerned with the finicky details that often
+make HTM transactions impractical, HTM does offer utility as a more flexible
+and performant synchronization primitive.
+
+### Software Value Prediction
+
+In TLS, there are some code regions which could be parallelized,
+but they involve potentially predictable data dependencies.
+For example, in a `while` loop, the iteration condition may not
+be known before executing the entire body of the loop and thus parallelism
+becomes very limited.
+
+Value prediction transforms this sequential execution into
+a potentially parallel one by creating data dependencies between
+the original variable and a prediction variable. Value prediction
+produces code with the following invariant:
+
+*Let x be some variable in the program, pred(x) is its predicted
+value and real(x) is its real value. A TLS task that reads pred(x)
+will be squashed whenever real(x) != pred(x).*
```

Comment:
Replace "is" with "be" (twice) for more a more grammatical-sounding first sentence?